gui/src/installer/step/mod.rs

```diff
@@ -111,6 +336,141 @@ fn bitcoind_default_address(network: &Network) -> String {
     }
 }
 
+/// Looks for bitcoind executable path and returns `None` if not found.
+fn bitcoind_exe_path() -> Option<PathBuf> {
+    which::which("bitcoind").ok()
+}
+
+/// Get available port that is valid for use by internal bitcoind.
+// Modified from https://github.com/RCasatta/bitcoind/blob/f047740d7d0af935ff7360cf77429c5f294cfd59/src/lib.rs#L435
+pub fn get_available_port() -> Result<u16, Error> {
+    // Perform multiple attempts to get a valid port.
+    for _ in 0..10 {
+        // Using 0 as port lets the system assign a port available.
+        let t = TcpListener::bind(("127.0.0.1", 0))
+            .map_err(|e| Error::CannotGetAvailablePort(e.to_string()))?;
+        let port = t
+            .local_addr()
+            .map(|s| s.port())
+            .map_err(|e| Error::CannotGetAvailablePort(e.to_string()))?;
+        if port_is_valid(&port) {
+            return Ok(port);
+        }
+    }
+    Err(Error::CannotGetAvailablePort(
+        "Exhausted attempts".to_string(),
+    ))
+}
+
+/// Checks if port is valid for use by internal bitcoind.
+pub fn port_is_valid(port: &u16) -> bool {
+    !BITCOIND_DEFAULT_PORTS.contains(port)
+}
+
+impl Default for SelectBitcoindTypeStep {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl From<SelectBitcoindTypeStep> for Box<dyn Step> {
+    fn from(s: SelectBitcoindTypeStep) -> Box<dyn Step> {
+        Box::new(s)
+    }
+}
+
+impl SelectBitcoindTypeStep {
+    pub fn new() -> Self {
+        Self {
+            use_internal: false,
+            error: None,
+        }
+    }
+}
+
+impl Step for SelectBitcoindTypeStep {
+    fn update(&mut self, message: Message) -> Command<Message> {
+        if let Message::SelectBitcoindType(msg) = message {
+            match msg {
+                message::SelectBitcoindTypeMsg::UseInternal(selected) => {
+                    self.use_internal = selected;
+                }
+            };
+            return Command::perform(async {}, |_| Message::Next);
+        };
+        Command::none()
+    }
+
+    fn apply(&mut self, ctx: &mut Context) -> bool {
```

Comment:
most of this code, should be;
in the StartBitcoindStep.load_context() where it populates StartBitcoindStep with information like the network.
then it should creates a command in  the step `load` method with a message returning the 
internal_bitcoind_config and internal_bitcoind_exe_config that the step will handle with the step `update` method`
the command will asynchonously take care of
- file reading like InternalBitcoindConfig::from_file
- `get_available_port` loop 
etc

For clarity you can wrap your step in an enum like:
```
pub enum InternalBitcoindStep {
   Saving{..} // state of the step while the command of the`load()` method is running
   Starting{..} // state of step starting the internal bitcoind step
   Started{..} // state of the step after the command of the start bitcoind is returned
}
```